# Artificial Neural Networks

### Customer retention Analytics

##### Business Problem:
An International bank collected a sample data of 10,000 customers. They observed that some of their custimers are leaving or churning in an unusually high rate and they want to find understand and access why their customers keep leaving. They have hiered you as a data scientist to look into the data to give then some insight.

##### Data:
The data was collected within the last 5 months, the feautures include the customer's name, creditscore, geography, gendey,age,tenure,balnce, number of products(accounts),credit card status(whether they have one or not), estimated salary, activity status(active member or not) and if the customer remained with them.

#### Goal:
Your goal is to create a geodemographic segmentation  to identify which of the customers have the highest risk of leaving.

## Import Libraries

In [1]:
import numpy as np
import pandas as  pd
import os
import matplotlib.pyplot as plt

# data encoding
from sklearn.preprocessing import LabelEncoder

# splitting the data
from sklearn.model_selection import train_test_split

# Standardization - feature scaling
from sklearn.preprocessing import StandardScaler

# Deep learning Libraries
import keras
from keras.models import Sequential
from keras.layers import Dense

# evaluating the mode
from sklearn.metrics import classification_report, confusion_matrix

Using TensorFlow backend.


In [ ]:
!python -m pip install --user --upgrade pip

In [ ]:
#DATA_DIR = 'data/'
#data = pd.read_csv(path + "Churn_Modelling.csv")

In [2]:
# os.getcwd()

# os.chdir('C:\\Users\\Wale\\Machine learning data')

# data = pd.read_csv('Churn_Modelling.csv')

'C:\\Users\\Wale'

In [5]:
data

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

## Data Preprocessing

In [6]:
X = data.iloc[:, 3:-1]
X

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619    France  Female   42       2       0.00              1   
1             608     Spain  Female   41       1   83807.86              1   
2             502    France  Female   42       8  159660.80              3   
3             699    France  Female   39       1       0.00              2   
4             850     Spain  Female   43       2  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771    France    Male   39       5       0.00              2   
9996          516    France    Male   35      10   57369.61              1   
9997          709    France  Female   36       7       0.00              1   
9998          772   Germany    Male   42       3   75075.31              2   
9999          792    France  Female   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  
0             1               1        101348.88  
1             0               1        112542.58  
2             1               0        113931.57  
3             0               0         93826.63  
4             1               1         79084.10  
...         ...             ...              ...  
9995          1               0         96270.64  
9996          1               1        101699.77  
9997          0               1         42085.58  
9998          1               0         92888.52  
9999          1               0         38190.78  

[10000 rows x 10 columns]

In [7]:
y = data.iloc[:,-1:]
y

Exited
0          1
1          0
2          1
3          0
4          0
...      ...
9995       0
9996       0
9997       1
9998       1
9999       0

[10000 rows x 1 columns]

#### Data encoding

In [8]:
# encoding country
encoder_X_1= LabelEncoder()
X.iloc[:,1] = encoder_X_1.fit_transform(X.iloc[:,1])

In [9]:
# encoding gender
encoder_X_2= LabelEncoder()
X.iloc[:,2] = encoder_X_2.fit_transform(X.iloc[:,2])

In [10]:
# we would also use the dummy variable because they are norminal variables


dummy = pd.get_dummies(X["Geography"], prefix = ['Geography'],drop_first=True)
X=pd.concat([X,dummy], axis = 1)
X.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0       0   42       2       0.00              1   
1          608          2       0   41       1   83807.86              1   
2          502          0       0   42       8  159660.80              3   
3          699          0       0   39       1       0.00              2   
4          850          2       0   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  ['Geography']_1  \
0          1               1        101348.88                0   
1          0               1        112542.58                0   
2          1               0        113931.57                0   
3          0               0         93826.63                0   
4          1               1         79084.10                0   

   ['Geography']_2  
0                0  
1                1  
2                0  
3                0  
4                1

In [11]:
X=X.drop(columns = ['Geography'], axis = 1)

In [12]:
X

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619       0   42       2       0.00              1          1   
1             608       0   41       1   83807.86              1          0   
2             502       0   42       8  159660.80              3          1   
3             699       0   39       1       0.00              2          0   
4             850       0   43       2  125510.82              1          1   
...           ...     ...  ...     ...        ...            ...        ...   
9995          771       1   39       5       0.00              2          1   
9996          516       1   35      10   57369.61              1          1   
9997          709       0   36       7       0.00              1          0   
9998          772       1   42       3   75075.31              2          1   
9999          792       0   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  ['Geography']_1  ['Geography']_2  
0                  1        101348.88                0                0  
1                  1        112542.58                0                1  
2                  0        113931.57                0                0  
3                  0         93826.63                0                0  
4                  1         79084.10                0                1  
...              ...              ...              ...              ...  
9995               0         96270.64                0                0  
9996               1        101699.77                0                0  
9997               1         42085.58                0                0  
9998               0         92888.52                1                0  
9999               0         38190.78                0                0  

[10000 rows x 11 columns]

## Install Kubeflow pipelines SDK

In [1]:
!pip3 install kfp --upgrade --user

Requirement already up-to-date: kfp in c:\users\wale\appdata\local\programs\python\python38\lib\site-packages (1.0.0)


You should consider upgrading via the 'c:\users\wale\appdata\local\programs\python\python38\python.exe -m pip install --upgrade pip' command.


Check if the install was successful:

In [2]:
!which dsl-compile

'which' is not recognized as an internal or external command,
operable program or batch file.


## Build the Components

In [ ]:
# Import Kubeflow SDK
import kfp
import kfp.dsl as dsl
import kfp.components as comp

In [ ]:
def train(data_path, model_file):
    
    # func_to_container_op requires packages to be imported inside of the function.
    import pickle
    
    import tensorflow as tf
    from tensorflow.python import keras
    
    # Download the dataset and split into training and test data. 
    DATA_DIR = 'data/'
    data = pd.read_csv(path + "Churn_Modelling.csv")

    X_train,X_test,y_train,y_test = train_test_split( X,y, test_size=0.2, random_state = 0)

    # Normalize the data.
    sc =StandardScaler()
    X_train = sc.fit_transform(X_train)
    X_test = sc.transform(X_test)

    # Define the model using Keras.
    classifier = keras.Sequential([
    keras.layers.Dense(output_dim =6,init='uniform', activation='relu', input_dim=11),
    keras.layers.Dense(output_dim =6,init='uniform', activation='relu'),
    keras.layers.Dense(output_dim =1,init='uniform', activation='sigmoid')])

    classifier.compile(optimizer = 'adam', loss='binary_crossentropy' 
                       , metrics =['accuracy'])
    
    # Run a training job with specified number of epochs
    classifier.fit(X_train, y_train, batch_size=10 , nb_epoch=100)

    # Evaluate the model and print the results
    test_loss, test_acc = model.evaluate(X_test,  y_test, verbose=0)
    print('Test accuracy:', test_acc)

    # Save the model to the designated 
    classifier.save(f'{data_path}/{model_file}')

    # Save the test_data as a pickle file to be used by the predict component.
    with open(f'{data_path}/test_data', 'wb') as f:
        pickle.dump((X_test,  y_test), f)

In [ ]:
def predict(data_path, model_file, churn_rate):
    
    # func_to_container_op requires packages to be imported inside of the function.
    import pickle

    import tensorflow as tf
    from tensorflow import keras

    import numpy as np
    
    # Load the saved Keras model
    classifier = keras.classifier.load_model(f'{data_path}/{model_file}')

    # Load and unpack the test_data
    with open(f'{data_path}/test_data','rb') as f:
        test_data = pickle.load(f)
    # Separate the test_images from the test_labels.
    X_test,  y_test = test_data


    # See https://github.com/kubeflow/pipelines/issues/2320 for explanation on this line.
    churn_rate = int(churn_rate)

    # make predictions.
    y_pred = classifier.predict(X_test)


    # create a threshold for the confution matrics
    y_pred=(y_pred>0.5)
    
    with open(f'{data_path}/result.txt', 'w') as result:
        result.write(" Prediction: {} ".format(y_pred))
    
    print('Prediction has be saved successfully!')

In [ ]:
# Create train and predict lightweight components.
train_op = comp.func_to_container_op(train, base_image='tensorflow/tensorflow:latest-gpu-py3')
predict_op = comp.func_to_container_op(predict, base_image='tensorflow/tensorflow:latest-gpu-py3')

## Build Kubeflow Pipeline

In [ ]:
#Create a client to enable communication with the Pipelines API server.
client = kfp.Client()

In [ ]:
# Define the pipeline
@dsl.pipeline(
   name='Customer Churn Pipeline',
   description='An ML pipeline that performs Customer Churn model training and prediction.'
)

# Define parameters to be fed into pipeline
def churn_container_pipeline(
    data_path: str,
    model_file: str, 
    churn_rate: int
):
    
    # Define volume to share data between components.
    vop = dsl.VolumeOp(
    name="create_volume",
    resource_name="data-volume", 
    size="1Gi", 
    modes=dsl.VOLUME_MODE_RWM)
    
    # Create MNIST training component.
    churn_training_container = train_op(data_path, model_file) \
                                    .add_pvolumes({data_path: vop.volume})

    # Create MNIST prediction component.
    churn_predict_container = predict_op(data_path, model_file, image_number) \
                                    .add_pvolumes({data_path: mnist_training_container.pvolume})
    
    # Print the result of the prediction
    mnist_result_container = dsl.ContainerOp(
        name="print_prediction",
        image='library/bash:4.4.23',
        pvolumes={data_path: churn_predict_container.pvolume},
        arguments=['cat', f'{data_path}/result.txt']
    )

In [ ]:
## Run the Pipeline

In [ ]:
DATA_PATH = '/churn'
MODEL_PATH='churn_classifier.h5'
# An integer representing an image from the test set that the model will attempt to predict the label for.
CHURN_RATE = 0

In [ ]:
pipeline_func = churn_container_pipeline

In [ ]:
experiment_name = 'costumer_churn_kubeflow'
run_name = pipeline_func.__name__ + ' run'

arguments = {"data_path":DATA_PATH,
             "model_file":MODEL_PATH,
             "image_number": CHURN_RATE}

# Compile pipeline to generate compressed YAML definition of the pipeline.
kfp.compiler.Compiler().compile(pipeline_func,  
  '{}.zip'.format(experiment_name))

# Submit pipeline directly from pipeline function
run_result = client.create_run_from_pipeline_func(pipeline_func, 
                                                  experiment_name=experiment_name, 
                                                  run_name=run_name, 
                                                  arguments=arguments)

In [21]:
# getting the predicted probabilities for each class
# These probabilities would help determine what customer has a high risk of leaving the bank
# y_pred = classifier.predict(X_test)
# y_pred

array([[0.20651075],
       [0.37878   ],
       [0.18120211],
       ...,
       [0.17714956],
       [0.17536563],
       [0.13352206]], dtype=float32)

From the results above we can see the probability of leaving the bank of a few of the customers. The bank can use this results by segmenting each customer by the risk of them leaving and develop a strategy to keep them.

In [22]:
# # create a threshold for the confution matrics
# y_pred=(y_pred>0.5)

In [23]:
# # confusion metrics
# cm = confusion_matrix(y_test, y_pred)
# print(cm)

[[1538   57]
 [ 261  144]]


From our confusion matrix we conclude that:
1. **True positive:** 144(We predicted a positive result and it was positive)- those  who left the bank over the five month period
2. **True negative:** 1538(We predicted a negative result and it was negative)-those  who didn't leave the bank over the five month periods
3. **False positive:** 57(We predicted a positive result and it was negative)-those who actually left but where predicted to have stayed.
4. **False negative:** 261(We predicted a negative result and it was positive)- those who were predicted to have left but didn't leave

In [24]:
#classification report
# cr = classification_report(y_test, y_pred)
# print(cr)

              precision    recall  f1-score   support

           0       0.85      0.96      0.91      1595
           1       0.72      0.36      0.48       405

    accuracy                           0.84      2000
   macro avg       0.79      0.66      0.69      2000
weighted avg       0.83      0.84      0.82      2000



The accuracy for the model might be really good but in this case what is important to us is the recall. We need to get the customers who churn correctly, which is the 1s. The recall for this class isn't very good so e might have to look first at the data imbalance and proceed from there to improve the results.